### PR TITLE
fix: .gitignore generated when running yarn init-test-app

### PR DESCRIPTION
**Pull request by Shivakumar-SG**
### Description

Added .gitignore files to `scripts/configure.js` and `/.npmignore`
 I actually merged both `example/.gitignore` and `example/windows/.gitignore` and added it into one. Which copies `.gitignore` files when running `yarn init-test-app`.
This solves issue #374 
Here is a sample screenshot :
Screenshot 2021-07-04 at 8 26 54 PM

### Platforms affected

- Android
- iOS
- macOS
- Windows

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -24,3 +24,5 @@ scripts/
 template-example/
 test/
 tsconfig.json
+example/.gitignore
+example/windows/.gitignore
```
Comment (tido64) [inline .npmignore]: It looks like you're excluding `.gitignore` files here. Did you mean to add `!` in front?
```suggestion
!example/.gitignore
!example/windows/.gitignore

```

You should also add an newline to remove the warning:

image

File: package.json [hunk 2]
```diff
@@ -95,6 +95,7 @@
     "react": "16.13.1",
     "react-native": "^0.63.4",
     "react-native-macos": "^0.63.33",
+    "react-native-test-app": "^0.6.10",
```
Comment (tido64) [inline package.json]: Why was this added? It shouldn't be necessary. Did you hit issues without it?

File: scripts/configure.js [hunk 3]
```diff
@@ -323,6 +323,10 @@ const getConfig = (() => {
             "metro.config.js": {
               source: path.join(testAppPath, "example", "metro.config.js"),
             },
+            ".gitignore": fs
+              .readFileSync(path.join(process.cwd(), "example", ".gitignore"))
```
Comment (tido64) [inline scripts/configure.js]: You shouldn't use current directory. We don't know where the user might be calling this script from. You should use `testAppPath` instead.

File: scripts/configure.js [hunk 4]
```diff
@@ -323,6 +323,10 @@ const getConfig = (() => {
...
+               + "\n" +
```
Comment (tido64) [inline scripts/configure.js]: I'm not sure you can just concatenate the two files. `windows/.gitignore` contains rules that may exclude names that are perfectly fine to have under the project folder. Can we instead copy the file under the Windows configuration section instead?

**Review by tido64 — CHANGES_REQUESTED** [review]
Thanks for taking the time to do this. I've left some comments. Let me know if you have questions.

File: .npmignore (hunk 1 above)
Comment (Shivakumar-SG) [reply .npmignore]: Yes, Sorry. I didn't know much about `.npmignore` and your feedback helped me. Thank you

File: package.json (hunk 2 above)
Comment (Shivakumar-SG) [reply package.json]: No, I didn't hit any issues without it and it was auto-generated for me. So I thought I'd leave it that way. I will remove it.

**Comment by Shivakumar-SG** [comment]
I have made the changes as suggested. Let me know if I have done things in the required way.

File: scripts/configure.js [hunk 5]
```diff
@@ -323,6 +323,10 @@ const getConfig = (() => {
             "metro.config.js": {
               source: path.join(testAppPath, "example", "metro.config.js"),
             },
+            ".gitignore": fs
+              .readFileSync(path.join(testAppPath, "example", ".gitignore"),{
+                encoding: "utf-8",
+              }),
```
Comment (tido64) [inline scripts/configure.js]: Since we're not making any changes to this file, we can just copy it:
```suggestion
            ".gitignore": {
               source: path.join(testAppPath, "example", ".gitignore"),
             },
```
Same below.

File: scripts/configure.js [hunk 6]
```diff
@@ -508,7 +512,12 @@ const getConfig = (() => {
           },
         },
         windows: {
-          files: {},
+          files: {
+            ".gitignorewindows":
```
Comment (tido64) [inline scripts/configure.js]: Why was this renamed? I don't think Git will pick it up if you do.
```suggestion
            ".gitignore":
```

File: yarn.lock [hunk 7]
```diff
@@ -7875,6 +7875,17 @@ react-native-macos@^0.63.33:
     use-subscription "^1.0.0"
     whatwg-fetch "^3.0.0"
 
+react-native-test-app@^0.6.10:
```
Comment (tido64) [inline yarn.lock]: Can you remove this as well?

**Review by tido64 — CHANGES_REQUESTED** [review]

File: .npmignore (hunk 1 above)
Comment (tido64) [reply .npmignore]: No worries. This file is a bit confusing. I forget to add things myself. 😄

File: scripts/configure.js (hunk 6 above)
Comment (Shivakumar-SG) [reply scripts/configure.js]: Oh, yea. This is all new to me so I just got confused and I feel stupid now.
Obviously, we have to have the name `.gitignore` in order to git to pick it up but it doesn't seem to copy contents from `example/windows/.gitignore`
What do you want me to do?

Comment (tido64) [reply scripts/configure.js]: I made the changes locally and I can see the two new files just fine. How are you testing it? Can you push your changes so we can take a closer look?

**Comment by Shivakumar-SG** [comment]
I did push the code. 
It seems to work fine when I select all platforms that we want to create a test app for. But If I select the platform as windows alone, it does not copy contents from `example/windows/.gitignore`

File: yarn.lock [hunk 8]
```diff
@@ -7875,6 +7875,16 @@ react-native-macos@^0.63.33:
...
+  version "0.6.10"
+  resolved "https://registry.npmjs.org/react-native-test-app/-/react-native-test-app-0.6.10.tgz#426670261c43d17a4a1e74419cba6f5691dce88f"
+  integrity sha512-QnG+A9oaf20qRCsmnWh2WYfebmu/kn8DGN3OAUaNRpLaIh6j9/t5HIaT+6cn5bEmNpELCPaN1hfgyl2KmtE3RA==
+  dependencies:
+    chalk "^4.1.0"
+    prompts "^2.4.0"
+    rimraf "^3.0.0"
+    semver "^7.3.5"
+    yargs "^16.0.0"
+
```
Comment (tido64) [inline yarn.lock]: Can you please revert this file, otherwise `yarn` will complain about an invalid lock file.

**Comment by tido64** [comment]
> It seems to work fine when I select all platforms that we want to create a test app for. But If I select the platform as windows alone, it does not copy contents from `example/windows/.gitignore`

Right, it looks like the common config is overwriting the windows one here: https://github.com/microsoft/react-native-test-app/blob/3f03bea56bf7d72ae573e256d2870641e68b949f/scripts/configure.js#L616

We can probably reverse it but there are a couple of things I'd like to have a look at before we do that. In the meantime, I think we can merge this if you address the remaining comments. 👍

**Comment by Shivakumar-SG** [comment]
> We can probably reverse it but there are a couple of things I'd like to have a look at before we do that. In the meantime, I think we can merge this if you address the remaining comments. 👍

I tried copying the `.gitignore` file into each platform's configuration section and in the window's section, I tried to concatenate both the `example/.gitignore` and `example/windows/.gitignore`, and its working
Here are the contents from the `.gitignore` created in this case:
Screenshot 2021-07-05 at 2 17 22 PM

And in other platforms too the `.gitignore` file is copied as required.
So is this right? or I just should 
>Can you please revert this file, otherwise yarn will complain about an invalid lock file.

address this issue and push the code?

**Comment by tido64** [comment]
> So is this right? or I just should
> 
> > Can you please revert this file, otherwise yarn will complain about an invalid lock file.
> 
> address this issue and push the code?

I don't think that will be right. The rules in `.gitignore` is relative to the file path. If we move them into each platform, they will not apply the rules in the project root. Let's address the one issue and I can take a closer look later.

File: yarn.lock (hunk 8 above)
Comment (tido64) [reply yarn.lock]: Sorry, I meant can we undo all changes within this file. This section was added because you added `react-native-test-app` to `package.json` earlier.

File: yarn.lock [hunk 9]
```diff
@@ -7875,6 +7875,7 @@ react-native-macos@^0.63.33:
     use-subscription "^1.0.0"
     whatwg-fetch "^3.0.0"
 
+
```
Comment (tido64) [inline yarn.lock]: I think you missed a line when reverting this file 😄 
```suggestion
```

File: scripts/configure.js [hunk 10]
```diff
@@ -508,7 +511,11 @@ const getConfig = (() => {
           },
         },
         windows: {
-          files: {},
+          files: {
+            ".gitignore": {
+               source: path.join(testAppPath, "example","windows", ".gitignore")
```
Comment (tido64) [inline scripts/configure.js]: Build is currently failing due to a formatting issue. Can you please run `yarn format:js`? That should fix it up.

image

**Review by tido64 — CHANGES_REQUESTED** [review]

**Review by tido64 — APPROVED** [review]
Looking good. Thanks for your hard work ❤️

**Comment by Shivakumar-SG** [comment]
Thanks for giving me the opportunity to do it. This is my first contribution to Open Source and I felt good doing it. Thank you so much @tido64 for bearing with me throughout sir.
